File: _eunjin/9934_완전_이진_트리.py
K = int(input())
arr = list(map(int, input().split()))

# 주어진 arr의 가운데 = root
# arr 왼쪽 부분 = 왼쪽 서브트리
# arr 오른쪽 부분 = 오른쪽 서브트리
# 서브트리 안에서의 가운데 = 그 서브트리의 root
# 서브트리로 좁혀갈 떄마다 depth +1

answer = [[] for _ in range(K)]

def recursion(start, end, depth):
    if start == end:
        answer[depth].append(arr[start])
        return

    root = (start + end) // 2
    recursion(start, root - 1, depth + 1)
    answer[depth].append(arr[root])
    recursion(root + 1, end, depth + 1)

recursion(0, len(arr) - 1, 0)

for row in answer:
    print(' '.join(map(str, row)))

File: _chongmyoung/9934_완전_이진_트리.java
import java.io.*;
import java.util.*;

public class Main {
    static int K;
    static int[] arr;
    static List<Integer>[] levels;

    // 중위 순회 방식으로 트리 구성
    static void buildTree(int start, int end, int depth) {
        if (start > end || depth >= K) return;

        int mid = (start + end) / 2;
        levels[depth].add(arr[mid]);

        buildTree(start, mid - 1, depth + 1);
        buildTree(mid + 1, end, depth + 1);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        K = Integer.parseInt(br.readLine());
        arr = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();
        levels = new ArrayList[K];
        for (int i = 0; i < K; i++) {
            levels[i] = new ArrayList<>();
        }

        buildTree(0, arr.length - 1, 0);

        // 각 레벨의 노드 출력
        for (int i = 0; i < K; i++) {
            System.out.println(String.join(" ", levels[i].stream()
                                                      .map(String::valueOf)
                                                      .toArray(String[]::new)));
        }
    }
}
